scvi/models/jvae.py

```diff
@@ -0,0 +1,372 @@
+from torch.nn import ModuleList
+
+from scvi.models.log_likelihood import log_zinb_positive, log_nb_positive
+from scvi.models.modules import Encoder
+from scvi.models.modules import MultiEncoder, MultiDecoder
+from scvi.models.utils import one_hot
+
+torch.backends.cudnn.benchmark = True
+
+
+class JVAE(nn.Module):
+    r"""Joint Variational auto-encoder
+
+    Implementation of gimVI:
+    *A joint model of unpaired data from scRNA-seq and spatial transcriptomics
+    for imputing missing gene expression measurements*
+    https://arxiv.org/abs/1905.02269
+
+    """
+
+    def __init__(
+        self,
+        dim_input_list: List[int],
+        total_genes: int,
+        indices_mappings: List[Union[np.ndarray, slice]],
+        reconstruction_losses: List[str],
+        model_library_bools: List[bool],
+        n_latent: int = 10,
+        n_layers_encoder_individual: int = 1,
+        n_layers_encoder_shared: int = 1,
+        dim_hidden_encoder: int = 128,
+        n_layers_decoder_individual: int = 0,
+        n_layers_decoder_shared: int = 0,
+        dim_hidden_decoder_individual: int = 32,
+        dim_hidden_decoder_shared: int = 128,
+        dropout_rate_encoder: float = 0.1,
+        dropout_rate_decoder: float = 0.3,
+        n_batch: int = 0,
+        n_labels: int = 0,
+        dispersion: str = "gene-batch",
+        log_variational: bool = True,
+    ):
+        """
+
+        :param dim_input_list: List of number of input genes for each dataset. If
+                the datasets have different sizes, the dataloader will loop on the
+                smallest until it reaches the size of the longest one
+        :param total_genes: Total number of different genes
+        :param indices_mappings: list of mapping the model inputs to the model output
+            Eg: [[0,2], [0,1,3,2]] means the first dataset has 2 genes that will be reconstructed at location [0,2]
+                                         the second dataset has 4 genes that will be reconstructed at [0,1,3,2]
+        :param reconstruction_losses: list of distributions to use in the generative process 'zinb', 'nb', 'poisson'
+        :param model_library_bools: bool list: model or not library size with a latent variable or use observed values
+        :param n_latent: dimension of latent space
+        :param n_layers_encoder_individual: number of individual layers in the encoder
+        :param n_layers_encoder_shared: number of shared layers in the encoder
+        :param dim_hidden_encoder: dimension of the hidden layers in the encoder
+        :param n_layers_decoder_individual: number of layers that are conditionally batchnormed in the encoder
+        :param n_layers_decoder_shared: number of shared layers in the decoder
+        :param dim_hidden_decoder_individual: dimension of the individual hidden layers in the decoder
+        :param dim_hidden_decoder_shared: dimension of the shared hidden layers in the decoder
```

Comment:
I think it could be a future improvement but I'm not sure it would really improve the model much compared to all the additional parameters it introduces. It would also be a broader change in the codebase, not only for GimVI.

I guess you should open an issue to change first `FCLayers` and eventually adapt the different models.

Reply:
Yep it's already there, another time then.